### PR TITLE
REP-110 regions

### PR DESCRIPTION
**Pull request by edwh**
This PR introduces the concept of regions, with businesses within them.  Initially we have two regions, London and Wales.

As discussed, the business repository is responsible for determining which businesses a regional admin can see using the location of the business and a polygon which defines the region.  This is done using MySQL spatial queries.

There's a migration script to set up the tables for the region; regional admins will be added to regions manually in the DB for now.

I've added a Region class for completeness though it's not used.

### CODE REVIEW

File: database/migrations/Version20210322125634.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('CREATE TABLE regions (uid INT AUTO_INCREMENT NOT NULL, name VARCHAR(255) NOT NULL, polygon geometry NOT NULL, PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB');
+        $this->addSql('CREATE TABLE users_regions (idusers_regions INT AUTO_INCREMENT NOT NULL, user INT(11) NOT NULL, region INT(11) NOT NULL, PRIMARY KEY(idusers_regions), UNIQUE INDEX users_region_unique_idx(user, region)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB');
```
Comment (ngm) [inline database/migrations/Version20210322125634.php:18]: Noting here that the id column is named idusers_regions - any particular reason?

Perhaps in retro we can have a brief chinwag about what we think DB naming conventions should be.

File: src/Domain/Models/Region.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * @package  TheRestartProject\RepairDirectory\Domain\Models
+ * @license  http://www.gnu.org/copyleft/gpl.html GNU General Public License
+ *
+ * @SuppressWarnings(PHPMD.ExcessivePublicCount)
+ * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
+ */
+class Region
+{
+    /**
+     * The business ID (generated by the infrastructure layer)
```
Comment (ngm) [inline src/Domain/Models/Region.php:21]: business -> region

File: src/Infrastructure/Doctrine/Repositories/DoctrineBusinessRepository.php [hunk 3]
```diff
@@ -38,29 +39,78 @@ public function add(Business $business)
...
+                INNER JOIN regions ON ST_Contains(regions.polygon, b.geolocation)
+                INNER JOIN users_regions ON regions.uid = users_regions.region
+                WHERE users_regions.user = " . intval($user->getUid()) . " AND b.uid = " . intval($uid);
+
+            $query = $this->entityManager->createNativeQuery(
+                $sql,
+                $rsm
+            );
+
+            $businesses = $query->getResult();
```
Comment (ngm) [inline src/Infrastructure/Doctrine/Repositories/DoctrineBusinessRepository.php:106]: Why do we get multiple businesses back if we're finding by id?

File: database/migrations/Version20210322125634.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+class Version20210322125634 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('CREATE TABLE regions (uid INT AUTO_INCREMENT NOT NULL, name VARCHAR(255) NOT NULL, polygon geometry NOT NULL, PRIMARY KEY(uid)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB');
```
[the remark this answers is not in the document]
Comment (edwh) [reply database/migrations/Version20210322125634.php:17]: I always use 'id', but as you say I'm being consistent with the other tables.

File: database/migrations/Version20210322125634.php (hunk 1 above)
Comment (edwh) [reply database/migrations/Version20210322125634.php:18]: I went for consistency with `idusers_groups` in the restarters `users_groups` table.

Personally I would use `id` here too, but...

File: src/Domain/Models/Region.php (hunk 2 above)
Comment (edwh) [reply src/Domain/Models/Region.php]: Thanks.

File: src/Infrastructure/Doctrine/Repositories/DoctrineBusinessRepository.php (hunk 3 above)
Comment (edwh) [reply src/Infrastructure/Doctrine/Repositories/DoctrineBusinessRepository.php:106]: The query always returns an array of results, even if there is only one.  We could probably use first() or something, but then we'd have to check whether any were found at all.  

The loop handles both the case of none found and one found, so I often do it this way.

**Pull request merged**